Add allocation_fallback_strategy option as fallback strategy for per-node strategy

lhp-nemlig opened the pull request:
**Description:** <Describe what has changed.>
<!--Ex. Fixing a bug - Describe the bug and how this fixes the issue.
Ex. Adding a feature - Explain what this achieves.-->

Adds allocation_fallback_strategy option as fallback strategy for per-node strategy, such that targets that are not related to a specific node can be assigned. 

The fallback strategy is configurable via the config option allocation_fallback_strategy

**Link to tracking Issue(s):** <Issue number if applicable>

- Resolves: #3477 

**Testing:** <Describe what testing was performed and which tests were added.>

Fixed per-node tests to expect non-node target to be assigned according to fallback strategy

**Documentation:** <Describe the documentation added.>

lhp-nemlig committed:
Add least-weighted as fallback strategy for per-node strategy

swiatekm commented on cmd/otel-allocator/allocation/strategy.go:
```diff
@@ -149,7 +149,7 @@ func init() {
 	}
 	err = Register(perNodeStrategyName, func(log logr.Logger, opts ...AllocationOption) Allocator {
-		return newAllocator(log, newPerNodeStrategy(), opts...)
+		return newAllocator(log, newPerNodeStrategy(newleastWeightedStrategy()), opts...)
```
I would much prefer the fallback strategy be `consistent-hashing`, and that should also be the default if we make this configurable.

The reason for this is that the `least-weighted` strategy isn't stable over time - the end result depends on what order changes to the set of targets arrive. To put it another way, with `least-weighted`, if you start multiple target allocator Pods, there's no guarantee that they give the same target allocation unless they're started at the exact same time. In comparison, `consistent-hashing` is stable, and you can scale the Target Allocator with it.

lhp-nemlig replied:
I changed the strategy to consistent-hashing

swiatekm reviewed (commented):
This is a breaking change, so we either have to mark it as such, or hide it behind a feature gate. I'd prefer the latter, but am willing to be convinced otherwise. We'll discuss it during today's SIG meeting, feel free to join if you'd like to contribute @lhp-nemlig.

On a more technical note, you'll need to sign the CLA and add a changelog entry before we can merge this PR.

lhp-nemlig committed:
Add changelog file

lhp-nemlig committed:
Change fallback strategy to consistent-hashing

lhp-nemlig committed:
Update changelog

lhp-nemlig committed:
Fix bad test condition that might pass even if target was not assigned

lhp-nemlig committed:
Make fallback strategy a config option

lhp-nemlig commented:
> This is a breaking change, so we either have to mark it as such, or hide it behind a feature gate. I'd prefer the latter, but am willing to be convinced otherwise. We'll discuss it during today's SIG meeting, feel free to join if you'd like to contribute @lhp-nemlig.
> 
> On a more technical note, you'll need to sign the CLA and add a changelog entry before we can merge this PR.

I have changed the PR to add fallback strategy as a config option, making this a non-breaking change as the default is NO fallback. 
That makes the changes a little more extensive of course, but I think I have implemented this in a mostly reasonable way :)

lhp-nemlig committed:
Update changelog

lhp-nemlig committed:
Merge branch 'main' into add_fallback_strategy_for_per_node_strategy

lhp-nemlig committed:
Add period to test comments

swiatekm commented:
> > This is a breaking change, so we either have to mark it as such, or hide it behind a feature gate. I'd prefer the latter, but am willing to be convinced otherwise. We'll discuss it during today's SIG meeting, feel free to join if you'd like to contribute @lhp-nemlig.
> > On a more technical note, you'll need to sign the CLA and add a changelog entry before we can merge this PR.
> 
> @swiatekm I have changed the PR to add fallback strategy as a config option, making this a non-breaking change as the default is NO fallback. That makes the changes a little more extensive of course, but I think I have implemented this in a mostly reasonable way :)

I'm ok with this, but keep in mind that in order for this to actually be usable, you'll also need to add an attribute to the OpenTelemetryCollector CRD. That's not a problem, but since in that case you're changing the public API, it's going to invite more scrutiny during review. Just FYI.

lhp-nemlig commented:
> > > This is a breaking change, so we either have to mark it as such, or hide it behind a feature gate. I'd prefer the latter, but am willing to be convinced otherwise. We'll discuss it during today's SIG meeting, feel free to join if you'd like to contribute @lhp-nemlig.
> > > On a more technical note, you'll need to sign the CLA and add a changelog entry before we can merge this PR.
> > 
> > 
> > @swiatekm I have changed the PR to add fallback strategy as a config option, making this a non-breaking change as the default is NO fallback. That makes the changes a little more extensive of course, but I think I have implemented this in a mostly reasonable way :)
> 
> I'm ok with this, but keep in mind that in order for this to actually be usable, you'll also need to add an attribute to the OpenTelemetryCollector CRD. That's not a problem, but since in that case you're changing the public API, it's going to invite more scrutiny during review. Just FYI.

Ok @swiatekm, I expect to join the SIG meeting and you can tell me what I need to do in order to complete this

swiatekm commented:
@lhp-nemlig as per our discussion during the SIG meeting, the simplest way to proceed with this change without needing to add attributes to the CRDs, would be to:

1. Keep your current changes making it configurable on the target allocator.
2. Add a feature flag which, when enabled, would set the fallback to the `consistent-hashing` strategy.

You can add feature flags [here](https://github.com/open-telemetry/opentelemetry-operator/blob/main/pkg/featuregate/featuregate.go) and the target allocator ConfigMap is generated [here](https://github.com/open-telemetry/opentelemetry-operator/blob/main/internal/manifests/targetallocator/configmap.go#L39). Feature flags are global, so you can just import the package, and add the config value for the fallback strategy if the flag is enabled. Does that make sense?

lhp-nemlig committed:
Merge branch 'main' into add_fallback_strategy_for_per_node_strategy

lhp-nemlig committed:
Add feature gate for enabling fallback strategy

lhp-nemlig commented:
@swiatekm Thanks for the clear instructions, I believe everything should be in order now.

lhp-nemlig committed:
Fix featuregate id

swiatekm commented on cmd/otel-allocator/allocation/per_node_test.go:
```diff
@@ -93,6 +103,78 @@ func TestAllocationPerNode(t *testing.T) {
+		// if third target, should be skipped
+		assert.True(t, found, "target with hash %s not found", item.Hash())
+
+		// only the first two targets should be allocated
```
This comment seems unnecessary. In this test, we expect all targets to be allocated.

lhp-nemlig replied:
Removed these 2 unnecessary comments

swiatekm commented on cmd/otel-allocator/main.go:
```diff
@@ -81,7 +81,7 @@ func main() {
 
 	allocatorPrehook = prehook.New(cfg.FilterStrategy, log)
-	allocator, err = allocation.New(cfg.AllocationStrategy, log, allocation.WithFilter(allocatorPrehook))
+	allocator, err = allocation.New(cfg.AllocationStrategy, log, allocation.WithFilter(allocatorPrehook), allocation.WithFallbackStrategy(cfg.AllocationFallbackStrategy))
```
I'd prefer if we only used allocation.WithFallbackStrategy here if it is actually set. Relying on the function to correctly deal with an empty strategy name feels unnecessarily error-prone.

lhp-nemlig replied:
Added conditional here

swiatekm commented on internal/manifests/targetallocator/configmap.go:
```diff
@@ -90,6 +90,11 @@ func ConfigMap(params Params) (*corev1.ConfigMap, error) {
 		taConfig["allocation_strategy"] = v1beta1.TargetAllocatorAllocationStrategyConsistentHashing
 	}
+
+	if featuregate.EnableTargetAllocatorFallbackStrategy.IsEnabled() {
+		taConfig["allocation_fallback_strategy"] = v1beta1.TargetAllocatorAllocationStrategyConsistentHashing
+	}
```
Could you check this in a unit test as well?

lhp-nemlig replied:
Added unit test for feature flag toggle

swiatekm commented on .chloggen/add_fallback_strategy_for_per_node_strategy.yaml:
```diff
@@ -0,0 +1,19 @@
+component: target allocator
+
+# A brief description of the change. Surround your text with quotes ("") if it needs to start with a backtick (`).
+note: Added allocation_fallback_strategy option as fallback strategy for per-node allocation strategy
+
+# One or more tracking issues related to the change
+issues: [3477]
+
+# (Optional) One or more lines of additional information to render under the primary note.
+# These lines will be padded with 2 spaces and then inserted directly into the document.
+# Use pipe (|) for multiline entries.
+subtext: |
+  If using per-node allocation strategy, targets that are not attached to a node will not
+  be allocated. As the per-node strategy is required when running as a daemonset, it is 
+  not possible to assign some targets under a daemonset deployment
```
Could you update the note and subtext with information about the feature flag?

lhp-nemlig replied:
Updated note and subtext

swiatekm reviewed (commented):
Looks good to me overall, had some minor comments and nitpicks. I'm fine merging this PR as-is when they're addressed, but in an ideal world we'd also include an e2e test showing the fallback strategy assigning targets without Nodes when the flag is enabled. I'm ok with that happening in a follow-up though, as this is an off-by-default feature flag, and writing these kinds of tests in chainsaw isn't very straightforward.

lhp-nemlig committed:
Update cmd/otel-allocator/allocation/per_node_test.go

Co-authored-by: Mikołaj Świątek <mail@mikolajswiatek.com>

lhp-nemlig committed:
Update cmd/otel-allocator/allocation/per_node_test.go

Co-authored-by: Mikołaj Świątek <mail@mikolajswiatek.com>

lhp-nemlig committed:
Update cmd/otel-allocator/allocation/per_node_test.go

Co-authored-by: Mikołaj Świątek <mail@mikolajswiatek.com>

lhp-nemlig committed:
Only add fallbackstrategy if nonempty

lhp-nemlig committed:
Remove unnecessary comments

lhp-nemlig committed:
Add unit test for fallbackstrategy feature gate

lhp-nemlig committed:
Update changelog

lhp-nemlig committed:
Merge branch 'main' into add_fallback_strategy_for_per_node_strategy

swiatekm reviewed (approved):
LGTM, thank you for the contribution!

jaronoff97 reviewed (approved):
Looks great! Thank you for your contribution :bow:

vidam-io commented:
Thanks for this feature! We're using it in production to scrape external VM metrics via ServiceMonitor with per-node allocation. Works great.
